bindings/py/cpp_src/bindings/encoders/py_SimHashDocumentEncoder.cpp

```diff
@@ -0,0 +1,189 @@
+/* -----------------------------------------------------------------------------
+ * HTM Community Edition of NuPIC
+ * Copyright (C) 2016, Numenta, Inc. https://numenta.com
+ *               2019, David McDougall
+ *               2019, Brev Patterson, Lux Rota LLC, https://luxrota.com
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU Affero Public License version 3 as published by
+ * the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU Affero Public License along with
+ * this program.  If not, see http://www.gnu.org/licenses.
+ * -------------------------------------------------------------------------- */
+
+/** @file
+ * py_SimHashDocumentEncoder.cpp
+ * @since 0.2.3
+ */
+
+#include <bindings/suppress_register.hpp>  //include before pybind11.h
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include <htm/encoders/SimHashDocumentEncoder.hpp>
+
+namespace py = pybind11;
+
+using namespace htm;
+
+
+namespace htm_ext {
+
+  using namespace htm;
+
+  void init_SimHashDocumentEncoder(py::module& m)
+  {
+    /**
+     * Parameters
+     */
+    py::class_<SimHashDocumentEncoderParameters>
+      py_SimHashDocumentEncoderParameters(m, "SimHashDocumentEncoderParameters",
+R"(
+Parameters for the SimHashDocumentEncoder.
+)");
+
+    py_SimHashDocumentEncoderParameters.def(py::init<>());
+
+    py_SimHashDocumentEncoderParameters.def_readwrite("activeBits",
+      &SimHashDocumentEncoderParameters::activeBits,
+R"(
+This is the number of true bits in the encoded output SDR. The output encoding
+will have a distribution of this many 1's. Specify only one of: activeBits
+or sparsity.
+)");
+
+    py_SimHashDocumentEncoderParameters.def_readwrite("size",
+      &SimHashDocumentEncoderParameters::size,
+R"(
+This is the total number of bits in the encoded output SDR.
+)");
+
+    py_SimHashDocumentEncoderParameters.def_readwrite("sparsity",
+      &SimHashDocumentEncoderParameters::sparsity,
+R"(
+This is an alternate way (percentage) to specify the the number of active bits.
+Specify only one of: activeBits or sparsity.
+)");
+
+    py_SimHashDocumentEncoderParameters.def_readwrite("tokenSimilarity",
+      &SimHashDocumentEncoderParameters::tokenSimilarity,
+R"(
+This allows similar tokens ("cat", "cats") to also be represented similarly,
+at the cost of document similarity accuracy. Default is FALSE (providing better
+document-level similarity, at the expense of token-level similarity).
+
+Results are heavily dependent on the content of your input data.
+
+If TRUE: Similar tokens ("cat", "cats") will have similar influence on the
```

Comment:
No, the plural form is the most obvious use-case, but there could be others.  Similarities here will just be things that are spelled nearly the same (or have the same chars in the strings). This param is not specific to any language.

Reply:
so this is implemented using `hamming distance==1: ignore those 'small' changes?`. If so, a) I would question its usefulness, as you get to ignore plurals, but also the most common single-letter typos. b) make the param an `UInt` instead: `minHammingDistance=2` (and differences below this value are ignored)

Reply:
Hamming distances are never considered or adjusted internally here, they're always the result of a kind of dynamic statistical distribution.

Here's a quick overview (needed something like this for a forum post, too). Do you think this would go well in my code documentation? Or too verbose?

Does this help you see what's going on?  This could work for plurals, mis-spellings, etc. I could also see it being hacked to create an advanced category encoder, where you can define the closeness and relation between categories, including in more than 1 dimension.

Let's say you have a document: **["hello", "there", "world"]**.

Tokens:
```
hello
there
world
```


## For `tokenSimilarity` OFF:

Hash Tokens:
```
hello => [ 0 1 0 1 0 1 0 1 1 ]
there => [ 0 0 0 1 0 0 1 0 0 ]
world => [ 1 0 0 1 0 1 0 1 0 ]
```

Prepare Token Hashes for Summation (convert 0 to -1):
```
hello => [ -1 +1 -1 +1 -1 +1 -1 +1 +1 ]
there => [ -1 -1 -1 +1 -1 -1 +1 -1 -1 ]
world => [ +1 -1 -1 +1 -1 +1 -1 +1 -1 ]
```

Document Summation:
```
sums  => [ -1 -1 -3 +3 -3 +1 -1 +1 -1 ]
```

Document SimHash (filter >= 0, or choose top N sums for sparse):
```
sim   => [  0  0  0  1  0  1  0  1  0 ]
```


## For `tokenSimilarity` ON:

Hash Tokens **and individual Characters**:
```
hello => [ 0 1 0 1 0 1 0 1 1 ]
    h => [ 1 0 1 1 0 1 0 0 1 ]
    e => [ 0 1 0 1 1 0 0 0 1 ]
    l => [ 1 1 0 1 1 0 1 0 0 ]
    l => [ 1 1 0 1 1 0 1 0 0 ]
    o => [ 1 1 1 0 0 0 0 0 0 ]
there => [ 0 0 0 1 0 0 1 0 0 ]
    t => [ 1 1 1 0 1 0 1 0 1 ]
    h => [ 1 0 1 1 0 1 0 0 1 ]
    e => [ 0 1 0 1 1 0 0 0 1 ]
    r => [ 0 0 0 1 1 0 0 0 0 ]
    e => [ 0 1 0 1 1 0 0 0 1 ]
world => [ 1 0 0 1 0 1 0 1 0 ]
    w => [ 0 0 0 0 1 1 1 0 1 ]
    o => [ 1 1 1 0 0 0 0 0 0 ]
    r => [ 0 0 1 1 0 0 0 1 1 ]
    l => [ 1 1 0 1 1 0 1 0 0 ]
    d => [ 0 1 0 0 1 0 0 0 0 ]
```

Prepare Token/Char Hashes for Summation (convert 0 to -1):
```
hello => [ -1 +1 -1 +1 -1 +1 -1 +1 +1 ]
    h => [ +1 -1 +1 +1 -1 +1 -1 -1 +1 ]
    e => [ -1 +1 -1 +1 +1 -1 -1 -1 +1 ]
    l => [ +1 +1 -1 +1 +1 -1 +1 -1 -1 ]
    l => [ +1 +1 -1 +1 +1 -1 +1 -1 -1 ]
    o => [ +1 +1 +1 -1 -1 -1 -1 -1 -1 ]
there => [ -1 -1 -1 +1 -1 -1 +1 -1 -1 ]
    t => [ +1 +1 +1 -1 +1 -1 +1 -1 +1 ]
    h => [ +1 -1 +1 +1 -1 +1 -1 -1 +1 ]
    e => [ -1 +1 -1 +1 +1 -1 -1 -1 +1 ]
    r => [ -1 -1 -1 +1 +1 -1 -1 -1 -1 ]
    e => [ -1 +1 -1 +1 +1 -1 -1 -1 +1 ]
world => [ +1 -1 -1 +1 -1 +1 -1 +1 -1 ]
    w => [ -1 -1 -1 -1 +1 +1 +1 -1 +1 ]
    o => [ +1 +1 +1 -1 -1 -1 -1 -1 -1 ]
    r => [ -1 -1 +1 +1 -1 -1 -1 +1 +1 ]
    l => [ +1 +1 -1 +1 +1 -1 +1 -1 -1 ]
    d => [ -1 +1 -1 -1 +1 -1 -1 -1 -1 ]
```

Document Summation:
```
sums  => [ +0 +2 -6 +8 -2 -8 -6 -12 +0 ]
```

Document SimHash (filter >= 0, or choose top N sums for sparse):
```
sim   => [  1  1  0  1  0  0  0  0  1 ]
```

Reply:
> Hamming distances are never considered or adjusted internally here, 

ok, this was an important bit for me to understand! 

> Do you think this would go well in my code documentation?

yes, this is perfect and I needed something like this to accompany the new concept in your encoder :100: 
I'd say a `README_SimHashDocument_encoder.md` (or doc/odt) placed somewhere (we don't have docs, so near the .hpp?) and refer it from the hpp/py docs. 

> Does this help at all?

a lot, thank you! 

> Hash Tokens:
hello => [ 0 1 0 1 0 1 0 1 1 ]

I still need more help on this one. I guess it's the LSH (local sim hash?). Could you explain that briefly in the document too, or add a link to a simple paper? For sure the doc should mention this phase is where the LSH is done. 
- I need to understand what a representation of "hello" and "h" means, so I can see what logically happens when we "sum representations by cols". 
- the similarity OFF seems clear to me (appart for the hashes above), sim ON is still a bit confusing, is it that "any pattern I want to consider 'ignored' or similar, I add as a row and encode it for the summation"? 
  - so if I wanted to ignore slipped double-letters, i'd add {wo, ow, or, ro, rl, lr, ...}? 
  - and if I wanted ignored 'arrrrg' anywhere in the text, id add it? (as a row as in sim ON) 

 > This could be used towards plurals, mis-spellings, etc. 

make this also a section of the doc, this is interesting and important.

Reply:
> > Hash Tokens:
> > hello => [ 0 1 0 1 0 1 0 1 1 ]
> 
> I still need more help on this one.

This is just a little example hashing function with 9-bit output (which I made up in my head on the fly).  If we used a real hash function, let's say ADLER32, then `hello => 110001011000000001000010101`. In my code, it's SHA3+SHAKE256.  (In the RDSE, it's Murmur3. etc.)

- So, we take our Things.
- We hash each Thing.
- We sum the bit columns from the hashes of the Things to get a "histogram" or statistical distribution of the bits.
- Filter those sums back to binary.
- You've now got a SimHash for the Things.

With `tokenSimilarity` off, we are hashing only full strings, so letters/characters do not exist, including in the output representation.  Tokens which may happen to share letters/characters will not have similar representations.

With `tokenSimilarity` on, we are _also_ hashing each letter/char of the strings.  Letters/Chars now exist, including in the output representation.  Tokens which share letters/chars may now also share bits and have similar representations!

> make this also a section of the doc, this is interesting and important.

👍 I'll make sure this is added with my other incoming dox changes.

Reply:
> hashing function with 9-bit output (which I made up in my head on the fly)

oh, I'm starting to understand it. So there's no special hashing (meaning a specialized for NLP hashing logic), the document fingerprint is based on statistical occurence of the words/tokens. So "dog cat cat cat c++17" would be mostly about cats (?) 

> In my code, it's SHA3+SHAKE256. (In the RDSE, it's Murmur3. etc.)

in that case, would it be possible to use already existing MurMur, and avoid the new dependency? (not too bad if not) 

Thanks a lot, this finaly helped me to understand (maybe you can pick something from that explanation for your doc)

Reply:
> oh, I'm starting to understand it. So there's no special hashing (meaning a specialized for NLP hashing logic), the document fingerprint is based on statistical occurence of the words/tokens. So "dog cat cat cat c++17" would be mostly about cats (?)

Yes! :metal: Exactly. I hadn't actually thought about it in that specific way before, thanks.

> in that case, would it be possible to use already existing MurMur, and avoid the new dependency? (not too bad if not)

Only for output SDR's that were 32-bits or 128-bits in size. For everything else (like the usual `n=400`), SHA3+SHAKE is the best option by far (but I'll keep an eye out).

> Thanks a lot, this finaly helped me to understand (maybe you can pick something from that explanation for your doc)

Absolutely, will do! thanks much.

Reply:
>  32-bits or 128-bits in size (the only digest sizes Murmur3 can output). For everything else (like the usual n=400), SHA3+SHAKE is the best option by far

this is low priority, but the other way then? is there any specific property of MurMur3, or we could remove it and RDSE would use sha+shake? 
CC @ctrl-z-9000-times what are the key functionality of murmur that RDSE needs, can this [switch to sha+shake] be done? 

> [on MurMur3] . When using 128-bits, the x86 and x64 versions do not produce the same values, as the algorithms are optimized for their respective platforms.

speaking of which, @brev please add (if not already similar) test for exact output. Simply make some test-case, record the output, and make the test compare exactly to that string. 
It would be interesting that the encoding is exactly the same for all platforms (OSX,Linux, Win) (CI will check that). For now we managed to get the whole framework deterministic, so starting with same inputs produces same bits (SDRs) for everybody. 

> With tokenSimilarity on, we are also hashing each letter/char of the strings. 

I still "don't trust" this scenario. Not necessary for this PR, but someone should research how/if that really performs? Do 1-letter typos have largest similarity with original word? 
I am afraid the "atomization to letters" can skew the output SDR: "I love you. I also loooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooove mc chicken. I like bananas. I hate raviolli. I like coffee. I love steak." 
- with tokenization, this doc would be about literally `o` (it's an extreme example)

Reply:
> is there any specific property of MurMur3, or we could remove it and RDSE would use sha+shake?

Speed, if I understand correctly. SHA3 is cryptographic (slow), MurMur3 is not.

> please add (if not already similar) test for exact output. 

will do.